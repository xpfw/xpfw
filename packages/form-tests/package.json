{
  "name": "@xpfw/form-tests",
  "version": "1.0.0-beta.12",
  "main": "dist",
  "typings": "dist",
  "license": "MIT",
  "peerDependencies": {
    "lodash-es": "^4.x.x",
    "moment": "2.x.x",
    "react": "^16.x.x",
    "@xpfw/form": "1.x.x"
  },
  "dependencies": {
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@types/jest": "^24.0.4",
    "@types/lodash": "^4.14.121",
    "@types/lodash-es": "^4.17.2",
    "@types/mockdate": "^2.0.0",
    "@types/react": "^16.8.3",
    "@types/react-test-renderer": "^16.8.1",
    "@types/socket.io-client": "^1.4.32",
    "lodash-es": "^4.17.11",
    "moment": "^2.24.0",
    "@xpfw/form": "^1.0.0-beta.12",
    "jest": "^24.1.0",
    "jest-cli": "^24.1.0",
    "mockdate": "^2.0.2",
    "react": "^16.8.1",
    "react-test-renderer": "^16.8.2",
    "rollup": "^1.1.2",
    "rollup-plugin-typescript2": "^0.19.2",
    "rollup-plugin-uglify": "^6.0.2",
    "ts-jest": "=23.10.5",
    "tslib": "^1.9.3",
    "typescript": "^3.3.3",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": false
      }
    },
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleNameMapper": {
      "preact-render-to-json": "preact-render-to-json",
      "preact": "preact",
      "react": "preact-compat",
      "react-dom": "preact-compat"
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
