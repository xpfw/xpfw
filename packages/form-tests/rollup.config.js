// rollup.config.js
import typescript from "rollup-plugin-typescript2"
import {uglify} from "rollup-plugin-uglify"
import { minify } from "uglify-es"

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'cjs'
  },
  plugins: [
    typescript(),
    // uglify({}, minify)
  ],
  external: [
    "lodash",
    "moment",
    "mockdate",
    "feathers-client",
    "resub",
    "resub-persist",
    "socket.io-client",
    "@xpfw/validate",
    "@xpfw/form-shared",
    "react",
    "react-dom",
    "preact-render-to-json"
  ]
};