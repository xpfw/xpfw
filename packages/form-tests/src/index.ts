import arrayTest from "./tests/array"
import booleanTest from "./tests/boolean"
import dateTest from "./tests/date"
import locationTest from "./tests/location"
import objectTest from "./tests/object"
import selectTest from "./tests/select"
import stringTest from "./tests/string"
import {
  ArrayField, BooleanField, DateField, LocationField, NameField,
  NumberField, ObjectField, PwField, SelectField
} from "./testUtil/schema"

export {
  booleanTest, stringTest, arrayTest, dateTest,
  locationTest, objectTest, selectTest,
  ArrayField, BooleanField, DateField, LocationField, NameField,
  NumberField, ObjectField, PwField, SelectField
}
