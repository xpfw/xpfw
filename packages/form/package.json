{
  "name": "@xpfw/form",
  "version": "1.0.0-beta.0",
  "main": "dist",
  "typings": "dist",
  "license": "MIT",
  "peerDependencies": {
    "ajv": "^6.x.x",
    "lodash-es": "^4.x.x",
    "mobx": "^5.x.x",
    "react": "^16.x.x"
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@types/ajv": "^1.0.0",
    "@types/jest": "^24.0.4",
    "@types/lodash-es": "^4.17.2",
    "@types/mockdate": "^2.0.0",
    "@types/react": "^16.8.3",
    "ajv": "^6.9.1",
    "jest": "^24.1.0",
    "jest-cli": "^24.1.0",
    "lodash": "^4.17.11",
    "lodash-es": "^4.17.11",
    "mobx": "^5.9.0",
    "react": "^16.8.1",
    "rollup": "^1.1.2",
    "rollup-plugin-typescript2": "^0.19.2",
    "rollup-plugin-uglify": "^6.0.2",
    "ts-jest": "=23.10.5",
    "tslib": "^1.9.3",
    "typescript": "^3.3.3",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleNameMapper": {
      "^lodash-es$": "<rootDir>/node_modules/lodash/index.js"
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
