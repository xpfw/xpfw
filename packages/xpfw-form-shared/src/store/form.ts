import {
  globals, IField, IFieldError, IForm, IFormError, iterateFields,
  validateField, validateForm, ValidationRegistry
} from "@xpfw/validate"
import { get, isFunction, isNil, isObject, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"
import FormErrorStore from "./error"

@AutoSubscribeStore
export class FormStore extends StoreBase implements IPersistableStore {
  public name = "formData"
  private formData: any = {}

  public getPropKeys() { return ["formData"] }

  public setValue(valuePath: string, value: any, fullTrigger: boolean = false) {
    set(this.formData, valuePath, value)
    this.trigger(fullTrigger ? undefined : valuePath)
  }

  @autoSubscribe
  public getValue(@key valuePath: string): any {
    return get(this.formData, valuePath)
  }

  public getFormData(form: IForm, prefix: string = "", method: string = "create") {
    const fullData: any = {}
    const formData = this.formData
    iterateFields(form, (field) => {
      let defaultValue = get(field, "validate.defaultValue")
      if (method === globals.Method.Find) {
        defaultValue = undefined
      }
      let value = get(formData,
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, defaultValue)
      let converter: any = get(field, "validate.convert")
      if (!isFunction(converter) && isObject(converter)) {
        converter = converter[method]
      }
      if (isFunction(converter)) {
        value = converter(value)
      }
      if (method !== globals.Method.Find || method === globals.Method.Find && !isNil(value))  {
        set(fullData, field.mapTo, value)
      }
    })
    return fullData
  }

  public copyDocToFormData(doc: any, form: IForm, prefix: string = "") {
    iterateFields(form, (field) => {
        this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, get(doc, field.mapTo))
    })
    this.trigger()
  }

  public async validateField(field: IField, form: IForm, method: string = "create",
                             prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`
    try {
      await validateField(get(this.formData, valuePath), field, {
        method,
        field,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public async validateForm(form: IForm, method: string = "create", prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${form.model}`
    try {
      await validateForm(this.getFormData(form, prefix), form, {
        method,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public resetForm(form: IForm, prefix: string = "")  {
    iterateFields(form, (field) => {
      this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, undefined)
    })
  }
}

export default new FormStore()
