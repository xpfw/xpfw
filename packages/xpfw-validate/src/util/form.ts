import { isObject } from "lodash"
import { IField, IForm } from "../typeDef"

const getFieldsFromForm: (form: IForm) => IField[] = (form: IForm) => {
  const fields: IField[] = []
  if (isObject(form)) {
    for (const section of form.sections) {
      for (const field of section.fields) {
        fields.push(field)
      }
    }
  }
  return fields
}

export { getFieldsFromForm }
