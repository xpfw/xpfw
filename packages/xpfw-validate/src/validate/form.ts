import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
