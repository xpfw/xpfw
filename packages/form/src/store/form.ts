import { get, set } from "lodash"
import { action, observable } from "mobx"
import { prependPrefix } from "../util/prefixMaker"

/**
 * Holds data relevant for Forms such as values and validation errors
 */
export class FormStoreClass {
  /**
   * MobX Observable containing all form data
   */
  @observable
  public formData: any = {}
  /**
   * MobX Observable containing all form errors
   */
  @observable
  public errors: any = {}
  /**
   * MobX Observable containing Loading states
   */
  @observable
  public loading: any = {}

  /**
   * Acquire a value from the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   * @param defaultValue value to return if it is not set in the form data
   */
  public getValue(valuePath?: string, prefix?: string, defaultValue?: any): any {
    return get(this.formData, prependPrefix(valuePath, prefix), defaultValue)
  }

  /**
   * Set a value in the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setValue(valuePath?: string, value?: any, prefix?: string) {
    set(this.formData, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getError(valuePath?: string, prefix?: string): any {
    return get(this.errors, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setError(valuePath?: string, value?: any, prefix?: string) {
    set(this.errors, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getLoading(valuePath?: string, prefix?: string): any {
    return get(this.loading, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setLoading(valuePath: string, value: boolean, prefix?: string) {
    set(this.loading, prependPrefix(valuePath, prefix), value)
  }
}

const FormStore = new FormStoreClass()
export default FormStore
