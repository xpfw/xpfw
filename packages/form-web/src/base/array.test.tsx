import { ComponentRegistry } from "@xpfw/form"
import { arrayTest } from "@xpfw/form-tests"
import ArrayField from "./array"
import TextField from "./text"

ComponentRegistry.registerComponent("array", ArrayField)
ComponentRegistry.registerComponent("string", TextField)

test("Array Field Test", arrayTest)
