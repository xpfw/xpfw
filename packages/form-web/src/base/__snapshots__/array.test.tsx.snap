// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array Field Test: got three values 1`] = `
<div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value=","
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value=",asdf "
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value="asdf"
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex center"
  >
    <a
      className="is-fullwidth is-info"
      onClick={[Function]}
    >
      Add
    </a>
  </div>
</div>
`;

exports[`Array Field Test: increased size to four 1`] = `
<div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value=","
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value=",asdf "
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value="asdf"
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value={null}
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex center"
  >
    <a
      className="is-fullwidth is-info"
      onClick={[Function]}
    >
      Add
    </a>
  </div>
</div>
`;

exports[`Array Field Test: removed second item 1`] = `
<div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value=","
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value="asdf"
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex flex1 center"
  >
    <input
      onChange={[Function]}
      type="text"
      value={null}
    />
    <a
      className="button is-warning  iconMargin"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "1.2rem",
        }
      }
    >
      Delete
    </a>
  </div>
  <div
    className="flex center"
  >
    <a
      className="is-fullwidth is-info"
      onClick={[Function]}
    >
      Add
    </a>
  </div>
</div>
`;

exports[`Array Field Test: unchanged render 1`] = `
<div>
  <div
    className="flex center"
  >
    <a
      className="is-fullwidth is-info"
      onClick={[Function]}
    >
      Add
    </a>
  </div>
</div>
`;
